File: leetcode-76.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        se = dict()
        sq = dict()
        for c in t:
            if c not in se:
                se[c] = 0
            se[c] += 1
        def check():
            for c in se:
                if c not in sq or se[c] > sq[c]:
                    return False
            return True
        l = 0
        ansl = -1
        len1 = 999999
        for i in range(len(s)):
            if s[i] not in sq:
                sq[s[i]] = 0
            sq[s[i]] += 1
            while check() and l <= i:
                if len1 > i-l+1:
                    len1 = i-l+1
                    ansl = l
                sq[s[l]] -= 1
                l += 1

        return '' if ansl == -1 else s[ansl: ansl+len1]


if __name__ == "__main__":
    so = Solution()
    print(so.minWindow('ab', 'a'))

File: leetcode/leetcode-76.cpp
/*
 *@File Name          :leetcode-76.cpp
 *@Author             :cc
 *@Version            :1.0.0
 *@Date               :2022/3/30 9:45
 *@Description        :
 *@Function List      :
 *@History            :
 */

#include "string"
#include "algorithm"
#include "iostream"
#include "vector"
#include "map"

using namespace std;

class Solution {
public:
    string minWindow(string s, string t) {
        map<char, int> mt;
        map<char, int> ms;
        string ans = "";
        for (auto &x: t)
            mt[x]++;
        int l = 0, r = 0;
        int n = s.size();
        int m = t.size();
        if (m > n) return ans;
        int num = 0;
        while (r < n) {
            ms[s[r]]++;
            if (ms[s[r]] <= mt[s[r]]) num++;
            r++;
            while (num == m && ms[s[l]] > mt[s[l]]) {
                ms[s[l]]--;
                l++;
            }
            if (num == m && (r - l < ans.length() || ans.empty())) {
                ans = s.substr(l, r - l);;
            }
        }
        return ans;
    }
};